### PR TITLE
Add basic integration/e2e tests.

### PR DESCRIPTION
**Pull request by jasonkuster**
This contribution adds the failsafe plugin in the examples directory and adds a WordCount test which for now runs using the DataflowPipelineRunner.

### CODE REVIEW

**Comment by jasonkuster** [comment]
R: @davorbonaci

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        "--project=apache-beam-testing",
+        "--runner=DataflowPipelineRunner",
+        "--stagingLocation=gs://apache-beam-testing-temp-storage/staging/" + jobName,
+        "--output=gs://apache-beam-testing-temp-storage/output/" + jobName + "/results",
+        "--workerLogLevelOverrides="
+        + "{\"com.google.cloud.dataflow.sdk.util.UploadIdResponseInterceptor\":\"DEBUG\"}"};
+
+    WordCountOptions options = PipelineOptionsFactory.fromArgs(args).withValidation()
+        .as(WordCountOptions.class);
+    Pipeline p = Pipeline.create(options);
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: This part duplicates the code from WordCount. Technically, it is testing the forked code, not the original one, which will get out of sync very soon. Can we do better?

**Comment by kennknowles** [comment]
Looks like you are getting bitten by checkstyle errors a bit.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+
+/**
+ * End-to-end tests of WordCount.
+ */
+@RunWith(JUnit4.class)
+public class WordCountIT extends BatchE2ETest {
```
Comment (kennknowles) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Please spell out `IntegrationTest`. I can't help but think Information Technology every time I see this :-)

etc

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    String[] args = {
+        "--jobName=" + jobName,
+        "--runner=com.google.cloud.dataflow.sdk.testing.TestDataflowPipelineRunner",
+        "--stagingLocation=" + testFileLocation + "staging/" + jobName,
+        "--output=" + testFileLocation + "output/" + jobName + "/results",
+        "--workerLogLevelOverrides="
+        + "{\"com.google.cloud.dataflow.sdk.util.UploadIdResponseInterceptor\":\"DEBUG\"}"};
+
+    WordCount.main(args);
+    PipelineResult result = TestDataflowPipelineRunner.getPipelineResultByJobName(jobName);
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Leave as-is for now. In the future, we'd want this to be runner-agnostic.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    String jobName = "wordcount-" + generateTestIdentifier() + "-prod";
+    String[] args = {
+        "--jobName=" + jobName,
+        "--runner=com.google.cloud.dataflow.sdk.testing.TestDataflowPipelineRunner",
+        "--stagingLocation=" + testFileLocation + "staging/" + jobName,
+        "--output=" + testFileLocation + "output/" + jobName + "/results",
+        "--workerLogLevelOverrides="
+        + "{\"com.google.cloud.dataflow.sdk.util.UploadIdResponseInterceptor\":\"DEBUG\"}"};
+
+    WordCount.main(args);
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Provide a builder for args.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  @Category(RunnableOnService.class)
+  public void testE2EWordCountOnDataflow() throws Exception {
+    parseTestOptions("testFileLocation");
+    String testFileLocation = testOptionsMap.get("testFileLocation");
+    if (!testFileLocation.endsWith("/")) {
+      testFileLocation = testFileLocation + "/";
+    }
+
+    String jobName = "wordcount-" + generateTestIdentifier() + "-prod";
+    String[] args = {
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Move to a common class; leave --output here.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * End-to-end tests of WordCount.
+ */
+@RunWith(JUnit4.class)
+public class WordCountIT extends BatchE2ETest {
+  @Test
+  @Category(RunnableOnService.class)
+  public void testE2EWordCountOnDataflow() throws Exception {
+    parseTestOptions("testFileLocation");
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Move all this code to a common base class.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+
+/**
+ * End-to-end tests of WordCount.
+ */
+@RunWith(JUnit4.class)
+public class WordCountIT extends BatchE2ETest {
+  @Test
+  @Category(RunnableOnService.class)
+  public void testE2EWordCountOnDataflow() throws Exception {
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Rename: testE2eWordCount

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import org.junit.runners.JUnit4;
+
+
+/**
+ * End-to-end tests of WordCount.
+ */
+@RunWith(JUnit4.class)
+public class WordCountIT extends BatchE2ETest {
+  @Test
+  @Category(RunnableOnService.class)
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Remove.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java (hunk 2 above)
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: This is a common pattern for the fail-safe plugin. Please keep as is.

File: pom.xml [hunk 9]
```diff
@@ -132,6 +132,25 @@
 
   <profiles>
     <profile>
+      <id>dev</id>
```
Comment (davorbonaci) [inline pom.xml]: I think this is not needed. I'd remove.

File: pom.xml [hunk 10]
```diff
@@ -248,6 +267,7 @@
             <parallel>${testParallelValue}</parallel>
             <threadCount>4</threadCount>
             <groups>${testGroups}</groups>
+            <skipTests>${skipUnitTests}</skipTests>
```
Comment (davorbonaci) [inline pom.xml]: I'd cut.

File: examples/java/pom.xml [hunk 11]
```diff
@@ -232,6 +232,25 @@
         <groupId>org.jacoco</groupId>
         <artifactId>jacoco-maven-plugin</artifactId>
       </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-failsafe-plugin</artifactId>
+        <version>2.19.1</version>
```
Comment (davorbonaci) [inline examples/java/pom.xml]: Move version to parent pom.

File: examples/java/pom.xml [hunk 12]
```diff
@@ -232,6 +232,25 @@
...
+        <configuration>
+          <skipTests>${skipIntegrationTests}</skipTests>
```
Comment (davorbonaci) [inline examples/java/pom.xml]: I'd cut.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/BatchE2ETest.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/BatchE2ETest.java]: Move out of examples.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/E2ETest.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
+/*
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/E2ETest.java]: Move out of examples.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/StreamingE2ETest.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/StreamingE2ETest.java]: Move out of examples.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java [hunk 16]
```diff
@@ -152,6 +161,7 @@ public void process(List<JobMessage> messages) {
       Throwables.propagateIfPossible(e);
       throw Throwables.propagate(e);
     }
+    RESULT_MAP.put(options.getJobName(), job);
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java]: Move above; do it as soon as you have both job name and job.

File: runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java [hunk 17]
```diff
@@ -55,6 +63,8 @@
 public class TestDataflowPipelineRunner extends PipelineRunner<DataflowPipelineJob> {
   private static final String TENTATIVE_COUNTER = "tentative";
   private static final Logger LOG = LoggerFactory.getLogger(TestDataflowPipelineRunner.class);
+  private static final Map<String, PipelineResult> RESULT_MAP =
```
Comment (davorbonaci) [inline runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java]: RESULT_MAP --> EXECUTION_RESULTS

File: runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java [hunk 18]
```diff
@@ -75,6 +85,37 @@ public static TestDataflowPipelineRunner fromOptions(
     return new TestDataflowPipelineRunner(dataflowOptions);
   }
 
+  /**
+   * @return String with a unique test identifier based on the current date, time, and a random int.
+   */
+  private static String generateTestIdentifier() {
```
Comment (davorbonaci) [inline runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java]: Simplify.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
+/*
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java]: Don't like the name `TestPipelineArgs`. We have `TestPipeline` concept already, but this is unrelated to that.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+  public void add(String key, String val) {
+    args.put(key, val);
+  }
+
+  /**
+   * Gets the job name argument.
+   *
+   * @return Name of the job to be run.
+   */
+  public String getJobName() {
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java]: Dataflow-specific; probably shouldn't be exposed as such.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+    return args.get("jobName");
+  }
+
+  /**
+   * Builds a string array of args which can be used by an end-to-end test.
+   *
+   * @return Test args for an end-to-end test.
+   */
+  public String[] build() {
+    ArrayList<String> optArrayList = new ArrayList<String>();
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java]: ArrayList<>()

Comment (jasonkuster) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java (hunk 20 above)
Comment (jasonkuster) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java (hunk 19 above)
Comment (jasonkuster) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipelineArgs.java]: Renamed to E2EArgs

File: runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java (hunk 17 above)
Comment (jasonkuster) [inline runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java]: Done.

File: runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java [hunk 22]
```diff
@@ -75,6 +85,37 @@ public static TestDataflowPipelineRunner fromOptions(
...
+    return now + random;
+  }
+
+  /**
+   * @return TestPipelineArgs for a Dataflow pipeline.
+   */
+  public static TestPipelineArgs createPipelineArgs() {
+    HashMap<String, String> opts = new HashMap<String, String>();
+    TestPipelineArgs.parseTestOptions("testFileLocation");
+    String fullClassName = new Exception().getStackTrace()[1].getClassName().toLowerCase();
```
Comment (jasonkuster) [inline runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java]: Pulled out into utility class.

File: runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java (hunk 18 above)
Comment (jasonkuster) [inline runners/google-cloud-dataflow-java/src/main/java/com/google/cloud/dataflow/sdk/testing/TestDataflowPipelineRunner.java]: Done.

File: examples/java/pom.xml [hunk 23]
```diff
@@ -232,6 +232,23 @@
...
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-failsafe-plugin</artifactId>
+        <configuration>
+          <useManifestOnlyJar>false</useManifestOnlyJar>
+          <redirectTestOutputToFile>true</redirectTestOutputToFile>
+        </configuration>
+        <executions>
+          <execution>
```
Comment (jasonkuster) [inline examples/java/pom.xml]: Adding the command line arg here.

File: examples/java/pom.xml [hunk 24]
```diff
@@ -232,6 +232,28 @@
...
+            <goals>
+              <goal>integration-test</goal>
+              <goal>verify</goal>
+            </goals>
+            <configuration>
+              <systemPropertyVariables>
+                <testOptions>testFileLocation=gs://clouddfe-testing-temp-storage</testOptions>
```
Comment (lukecwik) [inline examples/java/pom.xml]: We should make this part of the Jenkins configuration instead of having it explicitly here.

This will mirror the runnable on service setup within Jenkins.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/E2EArgs.java [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * {@link E2EArgs} is a set of args/utilities for use by Beam End-to-End tests.
+ */
+public class E2EArgs {
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/E2EArgs.java]: EndToEndTestArguments

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * End-to-end tests of WordCount.
+ */
+@RunWith(JUnit4.class)
+public class WordCountIT {
+  @Test
+  public void testE2EWordCount() throws Exception {
+    E2EArgs args = TestDataflowPipelineRunner.createPipelineArgs();
+    args.add("output", Joiner.on("/").join(new String[]{E2EArgs.getTestFileLocation(),
```
Comment (davorbonaci) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: Wrap differently; spacing around { and }.

File: examples/java/pom.xml [hunk 27]
```diff
@@ -232,6 +232,28 @@
         <groupId>org.jacoco</groupId>
         <artifactId>jacoco-maven-plugin</artifactId>
       </plugin>
+
+      <plugin>
```
Comment (davorbonaci) [inline examples/java/pom.xml]: Activation: to be discussed.
Build of which module: to be discussed.

File: examples/java/pom.xml (hunk 24 above)
Comment (lukecwik) [inline examples/java/pom.xml]: Use a json list, then you can use PipelineOptions and/or a library to do all the heavy lifting instead of parsing inside of E2EArgs.

File: examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+        "output", args.getArg("jobName"), "results"}));
+
+    WordCount.main(args.build());
+    PipelineResult result =
+        TestDataflowPipelineRunner.getPipelineResultByJobName(args.getArg("jobName"));
```
Comment (lukecwik) [inline examples/java/src/test/java/com/google/cloud/dataflow/examples/WordCountIT.java]: TestDataflowPipelineRunner throws an exception when the job fails.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/E2EArgs.java (hunk 25 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/E2EArgs.java]: Use jackson to do this for you instead of coding it by hand.

File: examples/java/pom.xml (hunk 23 above)
Comment (jasonkuster) [inline examples/java/pom.xml]: Done.
